### PR TITLE
XIVY-13150 implement notification actions

### PR DESCRIPTION
**Pull request by ivy-lgi**
Before:
image

After:
image

I noticed some issues we probably should discuss. Probably easier to explain in a meeting, but here are some notes:

Even though `WebNotification` contains a `List<WebNotificationAction>`, I assume there is exactly one `WebNotificationType.INFO` and zero or one `WebNotificationType.RUN` for now. If this changes, the implementation of the UI has to change as well. If this stays, maybe we should just provide two individual attributes for these?

In case of a list (at least for `WebNotificationType.RUN`) we could iterate over all of them and display a link/icon for each. However, I do not know what they do. All would have the same icon and the same tool tip. Another attribute like a description would be necessary for customization. To also make this language specific seems cumbersome. Maybe an `Enum` again with which the right CMS entry is acquired?

For new task notifications, I cannot easily determine whether a task has been completed or not. This means, I cannot hide the run link/icon if that is the case. Maybe we should add a method to `WebNotificationAction` that determines whether the action should be shown or not? Although, this will require information about the task to be present either in the `WebNotificationAction` or the `WebNotification` itself. 

This makes me think we should split `NotifcationAction` more precisely into something like:
```
> NotificationAction
  > InfoAction (one per Notification, in its own attribute)
  > RunnableAction (one or a list per Notification)
    > StartTaskAction
    > etc.
```

Another idea is to simply provide `WebNotification` with the `ITask` and/or `ICase`. I think this way we could completely remove the whole `NotificationAction` infrastructure again, as we then have access to all the links through that. At least for the current requirements this would be sufficient, or am I missing something?

### CODE REVIEW

File: dev-workflow-ui/webContent/includes/template/notifications.xhtml [hunk 1]
```diff
@@ -37,11 +37,18 @@
         <p:dataScroller id="notifications" lazy="true" styleClass="notification-scroll" value="#{notificationBean.dataModel}" var="notification" chunkSize="20" mode="inline">
           <div class="#{notification.style} notification-container">
             <div class="p-d-flex p-jc-between">
-              <h:outputText id="notificationMessage" value="#{notification.message}" escape="false"/>
+              <h:outputLink id="notificationMessage" value="#{notification.infoAction.link()}" styleClass="notification-message">
+                <h:outputText value="#{notification.message}"/>
```
Comment (weissreto) [inline dev-workflow-ui/webContent/includes/template/notifications.xhtml:41]: still use escape=false if the template wants to make something <b> or color=red. etc.

File: dev-workflow-ui/webContent/includes/template/notifications.xhtml [hunk 2]
```diff
@@ -37,11 +37,18 @@
         <p:dataScroller id="notifications" lazy="true" styleClass="notification-scroll" value="#{notificationBean.dataModel}" var="notification" chunkSize="20" mode="inline">
           <div class="#{notification.style} notification-container">
             <div class="p-d-flex p-jc-between">
-              <h:outputText id="notificationMessage" value="#{notification.message}" escape="false"/>
+              <h:outputLink id="notificationMessage" value="#{notification.infoAction.link()}" styleClass="notification-message">
```
Comment (weissreto) [inline dev-workflow-ui/webContent/includes/template/notifications.xhtml:40]: Also make it optional if no info action is here only render the outputText

File: dev-workflow-ui/webContent/includes/template/notifications.xhtml [hunk 3]
```diff
@@ -37,11 +37,18 @@
...
+                <h:outputText value="#{notification.message}"/>
+              </h:outputLink>
+              <p:tooltip for="notificationMessage" value="Info" />
 
-              <div class="p-d-flex">
+              <div class="p-d-flex notification-action-container">
```
Comment (weissreto) [inline dev-workflow-ui/webContent/includes/template/notifications.xhtml:45]: Maybe split container to mark as read and actions container. Also add more spacing between them.

File: dev-workflow-ui/webContent/includes/template/notifications.xhtml [hunk 4]
```diff
@@ -37,11 +37,18 @@
...
+                <h:outputLink id="notificationRunAction" value="#{notification.runAction.link()}" rendered="#{notification.runAction != null}">
```
Comment (weissreto) [inline dev-workflow-ui/webContent/includes/template/notifications.xhtml:46]: Iterate over all non info actions

File: dev-workflow-ui/webContent/includes/template/notifications.xhtml [hunk 5]
```diff
@@ -37,11 +37,18 @@
...
+                  <i class="notification-action si si-controls-play p-mr-1 " />
```
Comment (weissreto) [inline dev-workflow-ui/webContent/includes/template/notifications.xhtml:47]: get the css icon from the action itself. The action should be a DTO that delivers different CSS icons for different kind of actions

**Review by weissreto — CHANGES_REQUESTED** [review]

**Comment by weissreto** [comment]
> Before: image
> 
> After: image
> 
> I noticed some issues we probably should discuss. Probably easier to explain in a meeting, but here are some notes:
> 
> Even though `WebNotification` contains a `List<WebNotificationAction>`, I assume there is exactly one `WebNotificationType.INFO` and zero or one `WebNotificationType.RUN` for now. If this changes, the implementation of the UI has to change as well. If this stays, maybe we should just provide two individual attributes for these?
> 
> In case of a list (at least for `WebNotificationType.RUN`) we could iterate over all of them and display a link/icon for each. However, I do not know what they do. All would have the same icon and the same tool tip. Another attribute like a description would be necessary for customization. To also make this language specific seems cumbersome. Maybe an `Enum` again with which the right CMS entry is acquired?
> 
> For new task notifications, I cannot easily determine whether a task has been completed or not. This means, I cannot hide the run link/icon if that is the case. Maybe we should add a method to `WebNotificationAction` that determines whether the action should be shown or not? Although, this will require information about the task to be present either in the `WebNotificationAction` or the `WebNotification` itself.
> 
> This makes me think we should split `NotifcationAction` more precisely into something like:
> 
> ```
> > NotificationAction
>   > InfoAction (one per Notification, in its own attribute)
>   > RunnableAction (one or a list per Notification)
>     > StartTaskAction
>     > etc.
> ```
> 
> Another idea is to simply provide `WebNotification` with the `ITask` and/or `ICase`. I think this way we could completely remove the whole `NotificationAction` infrastructure again, as we then have access to all the links through that. At least for the current requirements this would be sufficient, or am I missing something?

Let's keep it simple. Plan a meeting with Marcel Prügel and Team WaWa for next week so that we can discuss their requirements. If they have more requirements (css icon, description, etc) we can implement them then.

The new-task event provider can decide if the task is available. We do not have to do that on the UI. If the task is already done the user gets redirected to the detail page of the task and message is displayed that it is already done. Good enough.

**Comment by ivy-lgi** [comment]
Look if multiple non info actions were present:
image

**Review by weissreto — APPROVED** [review]
Changes are good. 
On the UI I would like that actions are on the same column than other action independent of the mark as read button. Talk to @ivy-lli  how to do that. Can be implemented on another PR.

**Comment by alexsuter** [comment]
@ivy-lgi I think this has broken the core test suite engine-integration

**Pull request merged**